### PR TITLE
add user profile stats by god

### LINKED ISSUE #107: Add Profile Stats by God
When visiting a user's profile page, match history and a rank card are currently displayed. 
image

We now have stats data available in the Mongo stats_civs collection. These are currently being used to power the statistics pages.

- Write a new backend endpoint in `/api` that will query mongo for stats for an individual profile.
  - Consider building it to accept filters by current patch, all time, etc.
- On the front end, create a new component to display a player's god stats, matchups, maps, etc. Use shadcn components whenever possible. Tables could work better than graphs here. https://ui.shadcn.com/

Example:
image
https://aoe4world.com/players/707064

Definition of done:
- A user can visit their profile and view their stats for different major gods.

### PR DESCRIPTION
**Pull request by cishiv**
Changes:

- Adds `/api/data/builds` and `/api/data/game_modes` (avoids us having to maintain hard-coded values, and has the upshot of having the `game_modes` list used in the filter only include `game_modes` that have been played) - added cache control headers to this so we don't overfetch on each page reload.
- Adds `/api/stats/users/gods` + service
- Adds `majorGodNameToIndex` function
- Adds the `playerGodStats.tsx` component

How it works:
- We use the `Matches` collection and Mongo filters to pull back the appropriate Match info (all matches for a provided profileId, with some combination of civ, patch and major god) - we then manipulate the returned data to get w/l and play rates, combined across game modes.
- I also opted **not** to centralise the fetch logic in `profile.tsx` (like the other components), it looked pretty monolithic already.

How it looks:
image

Extensions:
- I'd probably add a "Frequently Played" table in the future, but didn't want this PR to be too big as it's already getting big.

Closes #107

### CODE REVIEW

**Comment by erin-fitzpatric** [comment]
This looks really good, I think play rate might be bugged though?
image

**Review by erin-fitzpatric — CHANGES_REQUESTED** [review]
Thanks for adding the builds and game_modes routes. I'm ok with it since there is a cache. I am slightly concerned about how well this will scale - we will eventually have 1 million + matches. Does it really make sense to scan this whole collection to find a game mode? 

Thanks for pulling the fetch logic out of profile.tsx - there's a lot of bloat due to some of the velocity of which we have been iterating, but this is definitely how it should be done long term.

I love the frequently played players feature if you want to pick that up next. I also would like a timeline to show change in rank over time.

@skspade - take a look at these changes for when you work on your profile styling overhaul PR again.

If you can fix the play %, I'll got ahead ad merge this.

File: src/app/api/stats/user/gods/route.ts
Comment (erin-fitzpatric) [inline src/app/api/stats/user/gods/route.ts:1]: I like how you organized this. Will be nice to keep "user" stat queries grouped together in the `stats/user` directory.

**Comment by cishiv** [comment]
> This looks really good, I think play rate might be bugged though? image

Ah good catch, will take a look after work tomorrow.

**Comment by cishiv** [comment]
> Thanks for adding the builds and game_modes routes. I'm ok with it since there is a cache. I am slightly concerned about how well this will scale - we will eventually have 1 million + matches. Does it really make sense to scan this whole collection to find a game mode?

Fair callout on the scalability - I think at 1M+ matches we'd probably want indexes on a couple of the fields (like profile_id) which would reduce the search space for distinct. We could also just use the hard coded match types tbh - it's not _that_ much of a UX loss.

**Comment by erin-fitzpatric** [comment]
Looking forward to getting this out there! Let me know once you have made the above updates and I'll merge it in. Have a few other pressing front end stories I could use help with too if you want to dig into the stats pages 👀

**Comment by cishiv** [comment]
@erin-fitzpatric looked into the play rate issue.

It's _technically_ not a bug. If you filter by a single god, the returned data will give you the play rate of that god out of all the results (which in this case, is just that god). Do you want it to be an absolute play rate? Would be a small change (e.g. if you filter by a specific god, give the play rate for that god out of all the gods). Just want to confirm if thats the behaviour you want.

**Comment by erin-fitzpatric** [comment]
> @erin-fitzpatric looked into the play rate issue.
> 
> It's _technically_ not a bug. If you filter by a single god, the returned data will give you the play rate of that god out of all the results (which in this case, is just that god). Do you want it to be an absolute play rate? Would be a small change (e.g. if you filter by a specific god, give the play rate for that god out of all the gods). Just want to confirm if thats the behaviour you want.

Ya I don't think it is useful to have 100% showing for every god. I would expect it to be the percentage I have played that civ compared to other picks.

**Comment by cishiv** [comment]
Okay cool - good sanity check. I've pushed the updated code.

No filters:
image

Filter by god:
image

I'll do a performance tuning pass if I make more changes to this code in different PRs.

**Comment by erin-fitzpatric** [comment]
This looks awesome, thanks for your work! I was going to go ahead and merge this if you feel good about it. 

Would be awesome to get some timeline graphs for player elo over time and other stuff like that if you would be interested. I also need some major help in the stats sections. I have greyed out options for "map" and "matchup" stats that I would like to get out there.

**Review by erin-fitzpatric — APPROVED** [review]

**Pull request merged**